### PR TITLE
Add stateful marshaller support to the CustomMarshallerAttributeAnalyzer

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Add support and tests for analyzing the required shapes for stateful marshallers.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 1]
```diff
@@ -456,4 +462,34 @@
   <data name="RequiresAllowUnsafeBlocksTitle" xml:space="preserve">
     <value>LibraryImportAttribute requires unsafe code.</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="StatefulMarshallerRequiresFromManagedDescription" xml:space="preserve">
+    <value>A stateful marshaller that supports a managed-to-unmanaged marshal mode must provide a 'FromManaged' instance method that takes the managed value as a parameter and returns 'void'.</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:466]: We have varied phrasing referring to different shapes/support in the descriptions - let's choose one. Examples I see:
1. supports a managed-to-unmanaged marshal mode
2. supports marshalling from managed to unmanaged
3. 'Managed to Unmanaged with Caller-Allocated Buffer' shape is used

I prefer (2) over (1), since 'managed-to-unmanaged marshal mode' could be interpreted as `MarshalMode.ManagedToUnmanaged*` instead. I think giving the message the specific `MarshalMode` is sufficient for tying it to what exactly the user specified to indicate 'supports marshalling from ... to ...'. (3) seems a bit out of place (maybe unless we give the shapes official names and link to doc with them), so I think it might be clearer to spell out as 'supports marshalling from ... to ... with a caller-allocated buffer'

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 2]
```diff
@@ -456,4 +462,34 @@
...
+    <value>A stateful marshaller that supports marshalling from managed to unmanaged must provide a 'FromManaged' instance method that takes the managed value as a parameter and returns 'void'.</value>
+  </data>
+  <data name="StatefulMarshallerRequiresFromManagedMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a one-parameter instance method named 'FromManaged' that takes a '{2}' as a parameter and returns 'void'</value>
+  </data>
+  <data name="StatefulMarshallerRequiresToUnmanagedDescription" xml:space="preserve">
+    <value>A stateful marshaller that supports marshalling from managed to unmanaged must provide a 'ToUnmanaged' instance method that takes no parameters and returns the 'unmanaged' type.</value>
+  </data>
+  <data name="StatefulMarshallerRequiresToUnmanagedMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a zero-parameter instance method named 'ToUnmanaged' returns the 'unmanaged' type for the marshaller</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:475]:
```suggestion
    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a zero-parameter instance method named 'ToUnmanaged' that returns the 'unmanaged' type for the marshaller</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx [hunk 3]
```diff
@@ -456,4 +462,34 @@
...
+    <value>A stateful marshaller that supports marshalling from unmanaged to managed must provide a 'FromUnmanaged' instance method that takes the unmanaged value as a parameter and returns 'void'.</value>
+  </data>
+  <data name="StatefulMarshallerRequiresFromUnmanagedMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a one-parameter instance method named 'FromUnmanaged' that takes the 'unmanaged' value as a parameter and returns 'void'</value>
+  </data>
+  <data name="StatefulMarshallerRequiresToManagedDescription" xml:space="preserve">
+    <value>A stateful marshaller that supports marshalling from unmanaged to managed must provide a 'ToManaged' instance method that takes no parameters and returns the managed type.</value>
+  </data>
+  <data name="StatefulMarshallerRequiresToManagedMessage" xml:space="preserve">
+    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a zero-parameter instance method named 'ToManaged' returns '{2}'</value>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Resources/Strings.resx:487]:
```suggestion
    <value>The type '{0}' specifies that it supports the '{1}' marshal mode for '{2}' but it does not provide a zero-parameter instance method named 'ToManaged' that returns '{2}'</value>
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeAnalyzer.cs [hunk 4]
```diff
@@ -836,10 +984,245 @@ private bool TryGetElementTypeFromSpanType(ITypeSymbol spanTypeMaybe, [NotNullWh
...
+                        if (!unmanagedType.IsUnmanagedType && !unmanagedType.IsStrictlyBlittable())
+                        {
+                            diagnosticReporter.CreateAndReportDiagnostic(UnmanagedTypeMustBeUnmanagedRule, methods.ToUnmanaged.ToDisplayString());
+                        }
+                    }
+
+                    if (ManualTypeMarshallingHelper.ModeUsesUnmanagedToManagedShape(mode))
+                    {
+                        // Verify the unmanaged types match unmanaged->managed shape
+                        if (unmanagedType is not null && !SymbolEqualityComparer.Default.Equals(unmanagedType, fromUnmanagedCandidates[0].Parameters[0].Type))
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeAnalyzer.cs:1190]: Store `fromUnmanagedCandidates[0]` as a local?

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs [hunk 5]
```diff
@@ -376,150 +376,7 @@ public static int BufferSize
             await VerifyCS.VerifyCodeFixAsync(
                 source,
                 fixedSource,
-                VerifyCS.Diagnostic(CallerAllocConstructorMustHaveBufferSizeRule).WithLocation(0).WithArguments("MarshallerType", "byte"));
-        }
-
-        [Fact]
-        public async Task ModeThatUsesManagedToUnmanagedShape_Missing_ConvertToUnmanagedMethod_Marshaller_DifferentDocument_ReportsDiagnostic()
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs:383]: Why are these deleted?

**Review by elinor-fung — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs (hunk 5 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs:383]: I think that was a bad merge. I'll make sure to bring these back
